jobs/schedulers.go

```diff
@@ -87,89 +98,91 @@ func (srv *JobServer) InitSchedulers() *Schedulers {
 			}
+		}
 
-			for {
-				timer := time.NewTimer(1 * time.Minute)
-				select {
-				case <-schedulers.stop:
-					mlog.Debug("Schedulers received stop signal.")
-					timer.Stop()
-					return
-				case now = <-timer.C:
-					cfg := schedulers.jobs.Config()
-
-					for idx, nextTime := range schedulers.nextRunTimes {
-						if nextTime == nil {
-							continue
-						}
+		for {
+			timer := time.NewTimer(1 * time.Minute)
+			select {
+			case <-schedulers.stop:
+				mlog.Debug("Schedulers received stop signal.")
+				timer.Stop()
+				return
+			case now = <-timer.C:
+				cfg := schedulers.jobs.Config()
 
-						if time.Now().After(*nextTime) {
-							scheduler := schedulers.schedulers[idx]
-							if scheduler != nil {
-								if schedulers.isLeader && scheduler.Enabled(cfg) {
-									if _, err := schedulers.scheduleJob(cfg, scheduler); err != nil {
-										mlog.Error("Failed to schedule job", mlog.String("scheduler", scheduler.Name()), mlog.Err(err))
-									} else {
-										schedulers.setNextRunTime(cfg, idx, now, true)
-									}
+				for idx, nextTime := range schedulers.nextRunTimes {
+					if nextTime == nil {
+						continue
+					}
+
+					if time.Now().After(*nextTime) {
+						scheduler := schedulers.schedulers[idx]
+						if scheduler != nil {
+							if schedulers.isLeader && scheduler.Enabled(cfg) {
+								if _, err := schedulers.scheduleJob(cfg, scheduler); err != nil {
+									mlog.Error("Failed to schedule job", mlog.String("scheduler", scheduler.Name()), mlog.Err(err))
+								} else {
+									schedulers.setNextRunTime(cfg, idx, now, true)
 								}
 							}
 						}
 					}
-				case newCfg := <-schedulers.configChanged:
-					for idx, scheduler := range schedulers.schedulers {
-						if !schedulers.isLeader || !scheduler.Enabled(newCfg) {
-							schedulers.nextRunTimes[idx] = nil
-						} else {
-							schedulers.setNextRunTime(newCfg, idx, now, false)
-						}
+				}
+			case newCfg := <-schedulers.configChanged:
+				for idx, scheduler := range schedulers.schedulers {
+					if !schedulers.isLeader || !scheduler.Enabled(newCfg) {
+						schedulers.nextRunTimes[idx] = nil
+					} else {
+						schedulers.setNextRunTime(newCfg, idx, now, false)
 					}
-				case isLeader := <-schedulers.clusterLeaderChanged:
-					for idx := range schedulers.schedulers {
-						schedulers.isLeader = isLeader
-						if !isLeader {
-							schedulers.nextRunTimes[idx] = nil
-						} else {
-							schedulers.setNextRunTime(schedulers.jobs.Config(), idx, now, false)
-						}
+				}
+			case isLeader := <-schedulers.clusterLeaderChanged:
+				for idx := range schedulers.schedulers {
+					schedulers.isLeader = isLeader
+					if !isLeader {
+						schedulers.nextRunTimes[idx] = nil
+					} else {
+						schedulers.setNextRunTime(schedulers.jobs.Config(), idx, now, false)
 					}
 				}
-				timer.Stop()
 			}
-		})
+			timer.Stop()
+		}
 	}()
 
-	return schedulers
+	schedulers.running = true
```

Comment:
What about racy access? Do we need a mutex here?

Reply:
I've decided to protect only the higher layer. This gets called by `JobServer.StartSchedulers()` which should call it under lock.
If it makes sense I can add a comment explaining that it's not safe for concurrent usage. Ideally I wanted to unexport all these but they are used in some test in the enterprise repository.

Reply:
Yeah, a comment would be helpful.